---
title: "1-Mapping and counting"
# bibliography: bib.bib
date: "`r format (Sys.time(), format='%Y-%m-%d %H:%M:%S %z')`"
output: 
  pdf_document:
    pandoc_args: 
      - '../../common.yaml'
      - --listings
    includes: 
      in_header: '../../preamble.tex'
    toc: true
    number_sections: true
    fig_width: 7
    fig_height: 6
    fig_caption: true
    df_print: tibble
---

```{r, echo = F}
library(knitr)
opts_chunk$set(tidy.opts=list(width.cutoff=80),tidy=TRUE)
```

\newpage

# Description

The Cell Ranger (v3.0.2) application (10x Genomics) was then used to demultiplex the BCL files into FASTQ files (cellranger mkfastq), to perform alignment (to Cell Ranger human genome references 3.0.0 GRCh38/Ensembl93), barcode filtering, UMI counting and to produce geneâ€”barcode matrices (cellranger count).

# CellRanger count from fastq

The script below was used to do the mapping and counting with Cellranger. 

```{bash eval=FALSE}
#!/usr/bin/env bash

	cellranger count	--id=${id} \
				--fastqs="${FastqDir}/${id}" \
				--transcriptome=$REF

```

Here, `$id`, `$FastqDir` and `$REF` are the sample ID, the directory containing fastq files and transcriptome reference directory (human genome version GRCh38 ensembl build 93). 

# System Configuration {#sec:3}

```{bash eval=TRUE}
uname -a
```

```{bash eval=FALSE}
cellranger
```
```
/home/qiang/cellranger-3.0.2/cellranger-cs/3.0.2/bin
cellranger  (3.0.2)
Copyright (c) 2019 10x Genomics, Inc.  All rights reserved.
-------------------------------------------------------------------------------

Usage:
    cellranger mkfastq

    cellranger count
    cellranger aggr
    cellranger reanalyze
    cellranger mat2csv

    cellranger mkgtf
    cellranger mkref

    cellranger vdj

    cellranger mkvdjref

    cellranger testrun
    cellranger upload
    cellranger sitecheck
```


